[pytorch][distributed] add function to get nccl version for error messages

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27068 [pytorch][distributed] add function to get nccl version for error messages**

Same as #26583. Closes https://github.com/pytorch/pytorch/issues/23549.

Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```

Original commit changeset: c5635ce89de1

Differential Revision: [D17639476](https://our.internmc.facebook.com/intern/diff/D17639476/)

rohan-varma committed:
[pytorch][distributed] add function to get nccl version for error messages

Original commit changeset: c5635ce89de1

Differential Revision: [D17639476](https://our.internmc.facebook.com/intern/diff/D17639476/)

[ghstack-poisoned]

pietern reviewed (approved):
Thanks, @rohan-varma.

rohan-varma committed:
Update on "[pytorch][distributed] add function to get nccl version for error messages"


Same as #26583. Closes https://github.com/pytorch/pytorch/issues/23549.

Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```

Original commit changeset: c5635ce89de1

Differential Revision: [D17639476](https://our.internmc.facebook.com/intern/diff/D17639476/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[pytorch][distributed] add function to get nccl version for error messages"


Same as #26583. Closes https://github.com/pytorch/pytorch/issues/23549.

Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```

Original commit changeset: c5635ce89de1

Differential Revision: [D17639476](https://our.internmc.facebook.com/intern/diff/D17639476/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot retest this please

rohan-varma committed:
Update on "[pytorch][distributed] add function to get nccl version for error messages"


Same as #26583. Closes https://github.com/pytorch/pytorch/issues/23549.

Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```

Original commit changeset: c5635ce89de1

Differential Revision: [D17639476](https://our.internmc.facebook.com/intern/diff/D17639476/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[pytorch][distributed] add function to get nccl version for error messages"


Same as #26583. Closes https://github.com/pytorch/pytorch/issues/23549.

Adds a function that uses `ncclGetVersion` from the NCCL API to retrieve the NCCL version. Converts it into a readable string, and is called in NCCL-related error messages to log the NCCL version. Hopefully this will help with debugging NCCL errors. 

Test plan:
1) Modify `C10D_NCCL_CHECK` in `NCCLUtils.hpp` to always error by setting `ncclResult_t error = ncclSystemError`
2)  force an NCCL error with script `test/simulate_nccl_errors.py`:
Start master node: `python test/simulate_nccl_errors.py localhost 9124 0 2`
Start other node: `python test/simulate_nccl_errors.py localhost 9124 1 2`
On the master node, should see the following error message w/NCCL version:
```
Traceback (most recent call last):
  File "simulate_nccl_errors.py", line 29, in <module>
    process_group.allreduce(torch.rand(10).cuda(rank)).wait()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:375, unhandled system error, NCCL version 2.4.8
```

Original commit changeset: c5635ce89de1

Differential Revision: [D17639476](https://our.internmc.facebook.com/intern/diff/D17639476/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/NCCLUtils.cpp:
```diff
@@ -0,0 +1,31 @@
+#include <c10d/NCCLUtils.hpp>
+#include <mutex>
```
Question: why do we need this header?

rohan-varma replied:
This allows us to access `std::call_once` and `std::once_flag`

mrshenli reviewed (approved):
The test failure looks irrelevant. This PR should be good to go.

```
01:57:35 LINK : fatal error LNK1102: out of memory
```
